flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultExecutionSlotAllocator.java

```diff
@@ -0,0 +1,248 @@
+	 * this method returns an empty collection to indicate no location preference.
+	 *
+	 * @return The preferred locations based in input streams, or an empty iterable,
+	 *         if there is no input-based preference.
+	 */
+	@VisibleForTesting
+	static CompletableFuture<Collection<TaskManagerLocation>> getPreferredLocationsBasedOnInputs(
+			ExecutionVertexID executionVertexId,
+			InputsLocationsRetriever inputsLocationsRetriever) {
+		CompletableFuture<Collection<TaskManagerLocation>> preferredLocations =
+				CompletableFuture.completedFuture(Collections.emptyList());
+
+		Collection<CompletableFuture<TaskManagerLocation>> locationsFutures = new ArrayList<>();
+
+		Collection<Collection<ExecutionVertexID>> allProducers =
+				inputsLocationsRetriever.getConsumedResultPartitionsProducers(executionVertexId);
+		for (Collection<ExecutionVertexID> producers : allProducers) {
+
+			for (ExecutionVertexID producer : producers) {
+				Optional<CompletableFuture<TaskManagerLocation>> optionalLocationFuture =
+						inputsLocationsRetriever.getTaskManagerLocation(producer);
+				optionalLocationFuture.ifPresent(
+						(locationFuture) -> locationsFutures.add(locationFuture));
+				// If the parallelism is large, wait for all futures coming back may cost a long time.
+				if (locationsFutures.size() > MAX_DISTINCT_LOCATIONS_TO_CONSIDER) {
+					locationsFutures.clear();
+					break;
+				}
+			}
+
+			CompletableFuture<Collection<TaskManagerLocation>> uniqueLocationsFuture =
+					FutureUtils.combineAll(locationsFutures).thenApply(
+							(locations) -> {
+								Set<TaskManagerLocation> uniqueLocations = new HashSet<>(locations);
+								return uniqueLocations;
+							});
+			preferredLocations = preferredLocations.thenCombine(
+					uniqueLocationsFuture,
+					(locationsOnOneEdge, locationsOnAnotherEdge) -> {
+						if ((!locationsOnOneEdge.isEmpty() && locationsOnAnotherEdge.size() > locationsOnOneEdge.size())
+								|| locationsOnAnotherEdge.isEmpty()) {
+							return locationsOnOneEdge;
+						} else {
+							return locationsOnAnotherEdge;
+						}
+					});
+			locationsFutures.clear();
+		}
+		return preferredLocations;
+	}
+
+	/**
+	 * Computes and returns a set with the prior allocation ids from all execution vertices scheduled together.
+	 *
+	 * @param executionVertexSchedulingRequirements contains the execution vertices which are scheduled together
+	 */
+	@VisibleForTesting
+	static Set<AllocationID> computeAllPriorAllocationIds(
+			Collection<ExecutionVertexSchedulingRequirements> executionVertexSchedulingRequirements) {
+		HashSet<AllocationID> allPreviousAllocationIds = new HashSet<>();
+		for (ExecutionVertexSchedulingRequirements schedulingRequirements : executionVertexSchedulingRequirements) {
+			AllocationID priorAllocation = schedulingRequirements.getPreviousAllocationId();
+			if (priorAllocation != null) {
+				allPreviousAllocationIds.add(priorAllocation);
+			}
+		}
+		return allPreviousAllocationIds;
```

Comment:
return executionVertexSchedulingRequirements
				.stream()
				.map(ExecutionVertexSchedulingRequirements::getPreviousAllocationId)
				.filter(Objects::nonNull)
				.collect(Collectors.toSet());
This could be simple.